File: python/course-repo/solution/04-concurrency-and-rate-limits/src/webhooks/send_bulk_demo.py
"""Rate-limit pain demo: fan out N deliveries using `demo-*` Activity IDs.

Used by Module 04 Section 2 (the "real rate limit on the receiver" demo). Same
fan-out shape as send_bulk, but a separate `demo-*` ID prefix so the leftover
in-flight Activities from this demo do not collide with the `bulk-*` Activities
in Sections 1, 3, and 4.
"""

import asyncio
import sys
from datetime import timedelta

from temporalio.client import Client

from .activities import deliver_webhook
from .shared import WEBHOOK_RECEIVER_URL, TASK_QUEUE, WebhookDelivery


async def main(count: int = 30) -> None:
    client = await Client.connect("localhost:7233")
    handles = []
    for i in range(count):
        h = await client.start_activity(
            deliver_webhook,
            args=[WebhookDelivery(
                url=WEBHOOK_RECEIVER_URL,
                payload={"event_id": f"demo_{i:03d}", "type": "demo_rate_limit"},
                event_id=f"demo_{i:03d}",
            )],
            id=f"demo-{i:03d}",
            task_queue=TASK_QUEUE,
            start_to_close_timeout=timedelta(seconds=30),
        )
        handles.append(h)
    await asyncio.gather(*(h.result() for h in handles))
    print(f"All {count} deliveries completed.")


if __name__ == "__main__":
    asyncio.run(main(int(sys.argv[1]) if len(sys.argv) > 1 else 30))

File: java/course-repo/solution/04-concurrency-and-rate-limits/src/main/java/webhooks/SendBulkDemo.java
package webhooks;

import io.temporal.client.ActivityClient;
import io.temporal.client.ActivityClientOptions;
import io.temporal.client.ActivityHandle;
import io.temporal.client.StartActivityOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Rate-limit pain demo: fan out N deliveries using demo-* activity IDs.
 *
 * Separate from SendBulk so leftover in-flight retries from this demo do not
 * collide with the bulk-* IDs used in sections 1, 3, and 4.
 */
public class SendBulkDemo {
  public static void main(String[] args) {
    int count = args.length > 0 ? Integer.parseInt(args[0]) : 60;

    WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
    ActivityClient client =
        ActivityClient.newInstance(
            service, ActivityClientOptions.newBuilder().setNamespace("default").build());

    List<ActivityHandle<Integer>> handles = new ArrayList<>();
    for (int i = 0; i < count; i++) {
      String eventId = String.format("demo_%03d", i);
      Map<String, Object> payload = new HashMap<>();
      payload.put("eventId", eventId);
      payload.put("type", "demo_rate_limit");
      WebhookDelivery req = new WebhookDelivery(Shared.WEBHOOK_RECEIVER_URL, payload, eventId);

      StartActivityOptions options =
          StartActivityOptions.newBuilder()
              .setId(String.format("demo-%03d", i))
              .setTaskQueue(Shared.TASK_QUEUE)
              .setStartToCloseTimeout(Duration.ofSeconds(30))
              .build();

      handles.add(
          client.start(WebhookActivities.class, WebhookActivities::deliverWebhook, options, req));
    }

    // Blocks: with the receiver rate-limited, these Activities keep retrying on
    // every 429 and never all complete. Press Ctrl+C after watching the 429s.
    for (ActivityHandle<Integer> handle : handles) {
      handle.getResult();
    }

    System.out.println("All " + count + " deliveries completed.");
    System.exit(0);
  }
}
